File: backend/app/schemas/replay.py
"""Pydantic contracts and response schemas for deterministic historical replay and comparison."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, FiniteFloat

from app.models.track import TrackState
from app.schemas.history import (
    HistoricalAlertItem,
    HistoricalDetectionItem,
    HistoricalThreatItem,
)


class ReplayFilter(BaseModel):
    track_ids: list[str] = Field(default_factory=list)
    sensor_ids: list[str] = Field(default_factory=list)
    classifications: list[str] = Field(default_factory=list)
    event_types: list[str] = Field(default_factory=list)


class ReplayRequest(BaseModel):
    start_time: datetime
    end_time: datetime
    step_interval_seconds: FiniteFloat = Field(default=1.0, gt=0.0, le=3600.0)
    filters: ReplayFilter = Field(default_factory=ReplayFilter)


class ReplayTrackState(BaseModel):
    track_id: str
    state: TrackState
    latitude: float
    longitude: float
    altitude: float | None = None
    velocity: float | None = None
    heading: float | None = None
    confidence: float
    classification: str | None = None
    source_count: int = 1


class ReplaySnapshot(BaseModel):
    replay_time: datetime
    step_index: int
    is_complete: bool
    active_tracks: list[ReplayTrackState] = Field(default_factory=list)
    recent_detections: list[HistoricalDetectionItem] = Field(default_factory=list)
    active_alerts: list[HistoricalAlertItem] = Field(default_factory=list)
    active_threats: list[HistoricalThreatItem] = Field(default_factory=list)
    metrics: dict[str, Any] = Field(default_factory=dict)


class ReplayStepRequest(BaseModel):
    start_time: datetime
    current_time: datetime
    end_time: datetime
    step_interval_seconds: FiniteFloat = Field(default=1.0, gt=0.0, le=3600.0)
    steps: int = Field(default=1, ge=1, le=1000)
    filters: ReplayFilter = Field(default_factory=ReplayFilter)


class ReplayComparisonRequest(BaseModel):
    request_1: ReplayRequest
    request_2: ReplayRequest


class ReplayComparisonReport(BaseModel):
    identical: bool
    total_detections_match: bool
    total_tracks_match: bool
    total_alerts_match: bool
    total_threats_match: bool
    detections_count_1: int
    detections_count_2: int
    tracks_count_1: int
    tracks_count_2: int
    alerts_count_1: int
    alerts_count_2: int
    threats_count_1: int
    threats_count_2: int
    differences: list[str] = Field(default_factory=list)

File: apps/operator/src/types/replay.ts
import { HistoricalAlertItem, HistoricalDetectionItem, HistoricalThreatItem } from './history';
import { TrackState } from './track';

export interface ReplayFilter {
  track_ids?: string[];
  sensor_ids?: string[];
  classifications?: string[];
}

export interface ReplayRequest {
  start_time: string;
  end_time: string;
  step_interval_seconds?: number;
  filters?: ReplayFilter;
}

export interface ReplayStepRequest {
  request: ReplayRequest;
  current_step?: number;
  steps_to_advance?: number;
}

export interface ReplayTrackState {
  track_id: string;
  state: TrackState;
  latitude: number;
  longitude: number;
  altitude?: number | null;
  velocity?: number | null;
  heading?: number | null;
  confidence: number;
  classification: string;
  source_count: number;
}

export interface ReplaySnapshot {
  replay_timestamp: string;
  step_index: number;
  is_complete: boolean;
  active_tracks: ReplayTrackState[];
  recent_detections: HistoricalDetectionItem[];
  active_alerts: HistoricalAlertItem[];
  active_threats: HistoricalThreatItem[];
}

export interface ReplayDifference {
  field: string;
  value_1: unknown;
  value_2: unknown;
  delta?: number | null;
}

export interface ReplayComparisonReport {
  identical: boolean;
  total_detections_match: boolean;
  total_tracks_match: boolean;
  total_alerts_match: boolean;
  total_threats_match: boolean;
  detections_count_1: number;
  detections_count_2: number;
  tracks_count_1: number;
  tracks_count_2: number;
  alerts_count_1: number;
  alerts_count_2: number;
  threats_count_1: number;
  threats_count_2: number;
  differences: ReplayDifference[];
}

export interface ReplayComparisonRequest {
  request_1: ReplayRequest;
  request_2: ReplayRequest;
}
